### PR TITLE
generator field constraints

### PR DESCRIPTION
**Pull request by narcoticfresh**
this PR adds the ability to specify additional (and completely optional) Contraints to fields.. The use case now is Email validation. This use case has been added to Showcase, showing a 'strict' validation as provided by Symfony..

### CODE REVIEW

**Comment by narcoticfresh** [comment]
This is WIP: I'm creating old school UnitTests for `JsonDefinition` and the fields there before we merge..

**Comment by narcoticfresh** [comment]
Wuhuu, 

> Test Coverage has increased to 60% (+10%).

File: src/Graviton/GeneratorBundle/Tests/Definition/DefinitionTest.php [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+     * @return void
+     */
+    public function testInvalidHandling()
+    {
+        $jsonDef = new JsonDefinition($this->invalidPath);
+    }
+    /**
+     * inexistent
+     *
+     * @expectedException \Exception
```
Comment (hairmare) [inline src/Graviton/GeneratorBundle/Tests/Definition/DefinitionTest.php]: This should expect ``Symfony\Component\HttpFoundation\File\Exception\FileNotFoundException`` at some point.

File: src/Graviton/GeneratorBundle/Tests/Definition/DefinitionTest.php [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+    /**
+     * invalid handling
+     *
+     * @expectedException \RuntimeException
+     *
+     * @return void
+     */
+    public function testInvalidHandling()
+    {
+        $jsonDef = new JsonDefinition($this->invalidPath);
```
Comment (hairmare) [inline src/Graviton/GeneratorBundle/Tests/Definition/DefinitionTest.php]: ``$jsonDef`` is unused and may be removed.

File: src/Graviton/GeneratorBundle/Tests/Definition/DefinitionTest.php [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+     * no id
+     *
+     * @expectedException \RuntimeException
+     *
+     * @return void
+     */
+    public function testNoId()
+    {
+        $jsonDef = new JsonDefinition($this->noIdPath);
+        $id = $jsonDef->getId();
```
Comment (hairmare) [inline src/Graviton/GeneratorBundle/Tests/Definition/DefinitionTest.php]: ``$id`` is unused and should be removed.

File: resources/definition/Showcase.json [hunk 4]
```diff
@@ -145,6 +147,20 @@
                 "translatable": true
             },
             {
+                "name": "emailField",
+                "type": "varchar",
+                "title": "eMail",
+                "description": "A field containing a EMail Constraint",
```
Comment (hairmare) [inline resources/definition/Showcase.json]: Is "EMail" how we write this? I'd rather have "Email", "email" or "e-mail".

**Comment by hairmare** [comment]
You might also want to consider creating some tests for non-object/non-hash json cases.

* json file containing simple string (ie. ``"Hello Hans"``).
* json file containing null (ie. ``null``).
* empty file as json file

**Comment by narcoticfresh** [comment]
OK; I corrected the unused vars and Exception type.. and added simple string json file..
We don't need to additionally test a file just containing `null` or `"Hello Hans"` as those would be invalid json (already tested).. I added one test case containing `["string"]`.. `[null]` would be the same test case..

I think that's enough..

As to
> Is "EMail" how we write this? 

This is not to our choosing - this is exactly the string Symfony expects.. This gives us the flexibility of specifying whatever (custom or built-in) `Constraint` classes that are needed and/or available..

**Comment by narcoticfresh** [comment]
@hairmare The travis build seems to fail on coverage upload, do you have any idea what can cause this? I assume it's just a temporary issue?

> The command "php ocular.phar code-coverage:upload --format=php-clover coverage.clover" exited with 1.

**Comment by hairmare** [comment]
@narcoticfresh ocular is what uploads our coverage info from travis results to scrutinizer-ci. It seems scrutinizer is having issues today :(

We should probably rewrite that line in .travis.yml to be non-faily.

File: resources/definition/Showcase.json [hunk 5]
```diff
@@ -145,6 +147,20 @@
                 "translatable": true
             },
             {
+                "name": "emailField",
+                "type": "varchar",
+                "title": "eMail",
```
Comment (hairmare) [inline resources/definition/Showcase.json]: This is one way to write it.

File: resources/definition/Showcase.json [hunk 6]
```diff
@@ -145,6 +147,20 @@
...
+                "description": "A field containing a EMail Constraint",
+                "constraints": [
+                    {
+                        "name": "Email",
```
Comment (hairmare) [inline resources/definition/Showcase.json]: This is the third? Should this one look like symfony EMail the most? I am probably not getting something?

File: src/Graviton/GeneratorBundle/Definition/JsonDefinition.php [hunk 7]
```diff
@@ -61,8 +62,8 @@ public function __construct($filename)
 
         $this->doc = json_decode(file_get_contents($this->filename));
 
-        if (empty($this->doc)) {
-            throw new \RuntimeException(sprintf('Could not load %s', $filename));
+        if (empty($this->doc) || !is_object($this->doc)) {
+            throw new FileNotFoundException(sprintf('Could not load %s', $filename));
```
Comment (hairmare) [inline src/Graviton/GeneratorBundle/Definition/JsonDefinition.php]: This line is now technically not true, the file could be there and invalid.

This all stems from the ``json_decode`` above.

Are you willing to switch to using [Filesystem](http://api.symfony.com/2.6/Symfony/Component/Filesystem/Filesystem.html) and [serializer](http://jmsyst.com/bundles/JMSSerializerBundle) in this PR or shall I write up a story? Those components would abstract away much of the file and low-level json handling. ``JsonDefinition``s would then only need to be concerned about their own schema.

File: src/Graviton/GeneratorBundle/Definition/JsonDefinitionField.php [hunk 8]
```diff
@@ -208,7 +209,11 @@ public function getClassName()
      */
     public function isClassType()
     {
-        return preg_match('/^class\:/', $this->def->type);
+        $ret = false;
+        if (preg_match('/^class\:/', $this->def->type) > 0) {
+            $ret = true;
+        }
+        return $ret;
```
Comment (hairmare) [inline src/Graviton/GeneratorBundle/Definition/JsonDefinitionField.php]: I'm not sure why you didn't write 
```php
return preg_match('/^class\:/', $this->def->type) > 0;
```
Not that it really matters...

File: src/Graviton/GeneratorBundle/README.md [hunk 9]
```diff
@@ -1,6 +1,6 @@
 # Graviton GeneratorBundle
 
-If you are prompted for a config type you should always choose xml
+If you are prompted for a config type you should always choose xml
```
Comment (hairmare) [inline src/Graviton/GeneratorBundle/README.md]: o__O

File: src/Graviton/GeneratorBundle/Tests/Definition/DefinitionElementTest.php [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+            'doctrineType' => '\Document\ShowcaseContactCode',
+            'serializerType' => '\Document\ShowcaseContactCode',
+            'isClassType' => true
+        );
+
+        $this->assertEquals($def, $field->getDefAsArray());
+    }
+
+    /**
+     * lcoal def
```
Comment (hairmare) [inline src/Graviton/GeneratorBundle/Tests/Definition/DefinitionElementTest.php]: tpyo

File: src/Graviton/GeneratorBundle/Tests/Definition/resources/test-string.json [hunk 11]
```diff
@@ -0,0 +1 @@
+["string"]
```
Comment (hairmare) [inline src/Graviton/GeneratorBundle/Tests/Definition/resources/test-string.json]: An even smaller "string" test would be just the following

```js
"string"
```
This is valid JSON, junst as the following line

```js
null
```

or even

```js
""
```

File: src/Graviton/GeneratorBundle/Definition/JsonDefinitionField.php [hunk 12]
```diff
@@ -222,6 +227,20 @@ public function getLength()
...
 
     /**
+     * Returns defined Constraints for this field
+     *
+     * @return array Constraints
+     */
+    public function getConstraints()
+    {
+        $ret = array();
+        if (isset($this->def->constraints) && is_array($this->def->constraints)) {
```
Comment (hairmare) [inline src/Graviton/GeneratorBundle/Definition/JsonDefinitionField.php]: Have you though about writing a JSON-Schema for the format at some point? That would make the ``is_array`` check obsolete and the code here nicer. Should we write a story on that?

**Comment by hairmare** [comment]
@narcoticfresh I just triggered a rebuild on travis. We'll see what happens, scrutinizer seems to be re.

File: src/Graviton/GeneratorBundle/Definition/JsonDefinitionField.php (hunk 8 above)
Comment (narcoticfresh) [inline src/Graviton/GeneratorBundle/Definition/JsonDefinitionField.php]: Just thought it's more readable like this..

File: src/Graviton/GeneratorBundle/Definition/JsonDefinitionField.php (hunk 12 above)
Comment (narcoticfresh) [inline src/Graviton/GeneratorBundle/Definition/JsonDefinitionField.php]: Yes, JSON schema should be a separate story

**Comment by narcoticfresh** [comment]
OK; if this runs through now, ready to merge..

File: src/Graviton/GeneratorBundle/Definition/JsonDefinitionField.php (hunk 12 above)
Comment (hairmare) [inline src/Graviton/GeneratorBundle/Definition/JsonDefinitionField.php]: Is [here](https://issue.swisscom.ch/browse/GRV-1454).

File: src/Graviton/GeneratorBundle/Definition/JsonDefinition.php (hunk 7 above)
Comment (hairmare) [inline src/Graviton/GeneratorBundle/Definition/JsonDefinition.php]: Tracking for this is in [GRV-1455](https://issue.swisscom.ch/browse/GRV-1455)

**Comment by hairmare** [comment]
There seem to have been some changes over on scrutinizer. Somehow a lot of issues have been marked as fixed with this build (and they seem to be of the false positive sort for what I've seen). Not sure if the code coverage change partially relate to this or not, the shouldn't since jsondefs were both large and completely untested.
